### PR TITLE
Define runtime compile time types

### LINKED ISSUE #15838: Unmanaged C++ library unloadability in C# example
It would be great if there was examples not only how to load/unload c# libraries to AssemblyContext, but also how to do it with unmanaged libraries (e.g. written in C, C++, C++/CLI). Maybe including what modifications needs to be done in unmanaged library's source code for this feature to work. Better documentation in this area would open gates for more users to hook existing unmanaged code (which may be in-use for old closed-source industrial protocols, and remastering it not worth the cost) to .NET platform. Unmanaged code unloadability is useful, then you don't trust unmanaged code's ability to do multithreading properly, but you know that single-threaded scenarios are well-tested, so in this case it's cheaper to load multiple times the same unmanaged library as a C# disposable class instance to get the job done.

I may add some possible examples to this post then I do more research here.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: f3283764-f15a-fa5f-5106-6f294c780f7b
* Version Independent ID: b03cedce-9074-9686-0b31-045b2b4c862d
* Content: [How to use and debug assembly unloadability in .NET Core](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability#feedback)
* Content Source: [docs/standard/assembly/unloadability.md](https://github.com/dotnet/docs/blob/master/docs/standard/assembly/unloadability.md)
* Product: **dotnet**
* GitHub Login: @janvorli
* Microsoft Alias: **janvorli**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #15838 

Several locations in the C# guide describe the differences between "runtime type" and "compile time type", but these were never defined clearly.

Add that description to the discussion of types and explain why it matters.

### CODE REVIEW

**Comment by BillWagner** [comment]
This Snippets 5000 failure is from a file where I'm removing code. There's no project there.

File: docs/csharp/programming-guide/types/index.md [hunk 1]
```diff
@@ -85,111 +78,101 @@ The following image shows value types and reference types in the CTS:
...
+It can be inconvenient to create a named type for simple sets of related values that you don't intend to store or pass outside method boundaries. You can create *anonymous types* for this purpose. For more information, see [Anonymous Types](../classes-and-structs/anonymous-types.md).
+
+Ordinary value types can't have a value of [null](../../language-reference/keywords/null.md). However, you can create nullable value types by affixing a `?` after the type. For example, `int?` is an `int` type that can also have the value [null](../../language-reference/keywords/null.md). Nullable value types are instances of the generic struct type <xref:System.Nullable%601?displayProperty=nameWithType>. Nullable value types are especially useful when you're passing data to and from databases in which numeric values might be null. For more information, see [Nullable value types](../../language-reference/builtin-types/nullable-value-types.md).
```
Comment (gewarren) [inline docs/csharp/programming-guide/types/index.md:149]:
```suggestion
Ordinary value types can't have a value of [null](../../language-reference/keywords/null.md). However, you can create nullable value types by appending a `?` after the type. For example, `int?` is an `int` type that can also have the value [null](../../language-reference/keywords/null.md). Nullable value types are instances of the generic struct type <xref:System.Nullable%601?displayProperty=nameWithType>. Nullable value types are especially useful when you're passing data to and from databases in which numeric values might be null. For more information, see [Nullable value types](../../language-reference/builtin-types/nullable-value-types.md).
```

File: docs/csharp/programming-guide/types/index.md [hunk 2]
```diff
@@ -85,111 +78,101 @@ The following image shows value types and reference types in the CTS:
...
+
+## Compile-time type and runtime type
+
+Any variable may have distinct compile time and runtime types. The *compile-time type* is the declared or inferred type of the variable in the source code. The *runtime type* is the type of the instance referred to by that variable. Often those two types are the same, as in the following example:
```
Comment (gewarren) [inline docs/csharp/programming-guide/types/index.md:153]:
```suggestion
A variable can have different compile-time and run-time types. The *compile-time type* is the declared or inferred type of the variable in the source code. The *run-time type* is the type of the instance referred to by that variable. Often those two types are the same, as in the following example:
```

See https://styleguides.azurewebsites.net/Styleguide/Read?id=2719&topicid=31304.

File: docs/csharp/programming-guide/types/index.md [hunk 3]
```diff
@@ -85,111 +78,101 @@ The following image shows value types and reference types in the CTS:
...
+
+:::code language="csharp" source="snippets/index/Program.cs" id="CompileTimeType":::
 
-Ordinary value types cannot have a value of [null](../../language-reference/keywords/null.md). However, you can create nullable value types by affixing a `?` after the type. For example, `int?` is an `int` type that can also have the value [null](../../language-reference/keywords/null.md). Nullable value types are instances of the generic struct type <xref:System.Nullable%601?displayProperty=nameWithType>. Nullable value types are especially useful when you are passing data to and from databases in which numeric values might be null. For more information, see [Nullable value types](../../language-reference/builtin-types/nullable-value-types.md).
+In other cases, the declared type is different, as shown in the following two examples:
```
Comment (gewarren) [inline docs/csharp/programming-guide/types/index.md:157]:
```suggestion
In other cases, the compile-time type is different, as shown in the following two examples:
```

File: docs/csharp/programming-guide/types/index.md [hunk 4]
```diff
@@ -85,111 +78,101 @@ The following image shows value types and reference types in the CTS:
...
+
+:::code language="csharp" source="snippets/index/Program.cs" id="RuntimeTypes":::
+
+In both of the preceding examples, the runtime type is a `string`. The compile-time type is `object` in the first line, and `IEnumerable<char>` in the second.
```
Comment (gewarren) [inline docs/csharp/programming-guide/types/index.md:161]:
```suggestion
In both of the preceding examples, the run-time type is a `string`. The compile-time type is `object` in the first line, and `IEnumerable<char>` in the second.
```

File: docs/csharp/programming-guide/types/index.md [hunk 5]
```diff
@@ -85,111 +78,101 @@ The following image shows value types and reference types in the CTS:
...
+
+It's important to understand when the compile-time type and the runtime type apply when the two types are different for a variable. The compile-time type determines all the actions taken by the compiler. These compiler actions include method call resolution, overload resolution, available implicit and explicit casts. The runtime type determines all actions that are resolved at runtime. These runtime actions include virtual method call dispatch, `is` expressions and `switch` expression evaluation, and other type testing APIs. Understand which type applies to which operation to better understand how your code interacts with types.
```
Comment (gewarren) [inline docs/csharp/programming-guide/types/index.md:163]:
```suggestion
If the two types are different for a variable, it's important to understand when the compile-time type and the run-time type apply. The compile-time type determines all the actions taken by the compiler. These compiler actions include method call resolution, overload resolution, and available implicit and explicit casts. The run-time type determines all actions that are resolved at run time. These run-time actions include dispatching virtual method calls, evaluating `is` and `switch` expressions, and other type testing APIs. To better understand how your code interacts with types, recognize which action applies to which type.
```

**Review by gewarren — APPROVED** [review]
Looks good. I left some suggestions for you to consider.

**Pull request merged**